Graphite UDP server gets ClassCastException

Hello

I'm trying to use Riemann as a proxy in front of Graphite. At the moment I'm just passing straight through to Graphite. Here is my config:

    (logging/init {:file "/var/log/riemann/riemann.log"})

    (let [host "0.0.0.0"]
      (graphite-server {:host host :protocol :tcp})
      (graphite-server {:host host :protocol :udp}))

    (streams
      (graphite {:host "127.0.0.1" :port 2005}))

The TCP server works fine, but the UDP server gives this error every time it gets a message:

    WARN [2015-02-05 15:25:21,698] defaultEventExecutorGroup-2-1 - riemann.transport.udp - UDP handler caught
    io.netty.handler.codec.DecoderException: java.lang.RuntimeException: Graphite server parse error (java.lang.ClassCastException: io.netty.channel.socket.DatagramPacket cannot be cast to java.lang.CharSequence): #<DatagramPacket DatagramPacket(/172.28.128.1:51160 => /0:0:0:0:0:0:0:0:2003, UnpooledUnsafeDirectByteBuf(ridx: 0, widx: 2048, cap: 2048))>
            at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:99)
            at riemann.transport.graphite.proxy$io.netty.handler.codec.MessageToMessageDecoder$ff19274a.channelRead(Unknown Source)
            at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:333)
            at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:319)
            at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
            at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:333)
            at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:319)
            at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:166)
            at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:333)
            at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:319)
            at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:787)
            at io.netty.channel.nio.AbstractNioMessageChannel$NioMessageUnsafe.read(AbstractNioMessageChannel.java:93)
            at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:511)
            at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:468)
            at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:382)
            at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:354)
            at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:116)
            at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137)
            at java.lang.Thread.run(Thread.java:745)
    Caused by: java.lang.RuntimeException: Graphite server parse error (java.lang.ClassCastException: io.netty.channel.socket.DatagramPacket cannot be cast to java.lang.CharSequence): #<DatagramPacket DatagramPacket(/172.28.128.1:51160 => /0:0:0:0:0:0:0:0:2003, UnpooledUnsafeDirectByteBuf(ridx: 0, widx: 2048, cap: 2048))>
            at riemann.transport.graphite$graphite_frame_decoder$fn__5167.invoke(graphite.clj:89)
            at riemann.transport.graphite.proxy$io.netty.handler.codec.MessageToMessageDecoder$ff19274a.decode(Unknown Source)
            at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:89)
            ... 18 more

Any idea what is going wrong?

Thanks
-Ben

--- a/src/riemann/transport/graphite.clj
+++ b/src/riemann/transport/graphite.clj
@@ -74,13 +74,15 @@
   name, tags) from; to fill in default states or TTLs, and so on.
 
   If parser-fn is nil, defaults to identity."
-  [parser-fn]
+  [parser-fn protocol]
   (let [parser-fn (or parser-fn identity)]
     (proxy [MessageToMessageDecoder] []
       (decode [context message out]
         (try+
           (.add out
-                (-> message
+                (-> (if (= protocol :udp)
+                      (.toString (.content message) CharsetUtil/UTF_8)
+                      message)
                     decode-graphite-line
                     parser-fn))
           (catch Object e
@@ -126,7 +128,7 @@
              ^:shared string-encoder (StringEncoder.
                                        CharsetUtil/UTF_8)
              ^:shared graphite-decoder (graphite-frame-decoder
-                                         (:parser-fn opts))
+                                         (:parser-fn opts) protocol)
              ^{:shared true :executor shared-event-executor} handler
              graphite-message-handler)]
        (server (merge opts

--- a/test/riemann/transport/graphite_test.clj
+++ b/test/riemann/transport/graphite_test.clj
@@ -65,3 +65,31 @@
                          :ttl         nil}))))
         (finally
           (core/stop! core))))))
+
+(deftest round-trip-test-udp
+  (riemann.logging/suppress ["riemann.transport"
+                             "riemann.pubsub"
+                             "riemann.graphite"
+                             "riemann.core"]
+         (let [server (graphite-server {:pool-size 1 :block-start true :protocol :udp})
+          sink   (promise)
+          core   (core/transition! (core/core)
+                                   {:services [server]
+                                    :streams  [(partial deliver sink)]})]
+      (try
+        ; Open a client and send an event
+        (let [client (client/graphite {:protocol :udp})]
+          (client {:host "computar"
+                   :service "hi there" :metric 2.5 :time 123 :ttl 10})
+          ; Verify event arrives
+          (is (= (deref sink 1000 :timed-out)
+                 (event {:host        nil
+                         :service     "computar.hi.there"
+                         :state       nil
+                         :description nil
+                         :metric      2.5
+                         :tags        nil
+                         :time        123
+                         :ttl         nil}))))
+        (finally
+          (core/stop! core))))))

